### ISSUE TITLE
Talkback borders not drawn when FlutterView#enableTransparentBackground called

### ISSUE THREAD

**Issue by mklim** [flutter/flutter#37025]
Something about [`FlutterVIew#enableTransparentBackground`](https://github.com/flutter/engine/blob/9cf13568e2aa1f9ab37445037e2ef54ecb5146f8/shell/platform/android/io/flutter/view/FlutterView.java#L318-L326) isn't working with Talkback. My current guess is `setZOrderOnTop(true)` may be setting the `SurfaceTexture` to render _on top of_ however TalkBack tries to draw its UI. That would be extremely unexpected behavior though.

Still not sure if this is an underlying issue with Flutter or Talkback. Going to try to produce the bug with a non Flutter app next.

This is probably related to #33811.

# Reproduction steps.

1. `flutter create foo && cd foo`
2. Add `getFlutterView().enableTransparentBackground();` to `onCreate` in `android/app/src/main/java/com/example/foo/MainActivity.java`
3. `flutter run` to an Android device with Talkback enabled
4. Observe app deploys successfully.
5. Tap on any UI element.

**Expected**: Talkback announces the tapped UI element and highlights it with a green box.
**Actual**: Talkback only announces the tapped UI element, and no highlighted green box is visible.

**Comment by mklim** [comment]
Removing `setZOrderOnTop(true)` from `enableTransparentBackground()` "fixes" this.

**Comment by mklim** [comment]
Some discussion offline. So the `setZOrderOnTop` API docs mention _By setting this, you cause it to be placed above the window. This means that none of the contents of the window this SurfaceView is in will be visible on top of its surface_. Apparently the accessibility highlights are also "contents of the window." So calling that API causes the Flutter UI to hide Talkback's highlights.

You can sort of see this by taking the reproduction app above and setting `backgroundColor: Colors.transparent`. The Talkback highlights will show up in places over the "white" (transparent) background but be hidden by opaque UI elements like the title bar.

There isn't an immediately obvious fix or workaround here, unfortunately. The `setZOrderOnTop` call is necessary for the SurfaceView to end up positioned on top of Android app UI elements so that the Flutter UI can be transparent and blend with them instead of blending to black. Maybe The API should be deprecated because of this bug, and people should be encouraged to use `FlutterTextureView` instead for this purpose. But that does necessitate migrating to the new embedding API.

Not sure if it makes sense for #33811 to be calling the same or similar API somewhere and ending up with the same root cause. /cc @matthew-carroll

**Comment by kf6gpe** [comment]
We believe this may be a platform issue (the `SurfaceView`); both Flutter team and the `customer: fun (g3)` team will reach out to the Android team as well.

**Comment by mklim** [comment]
`customer: fun` has been able to work around this by calling [`setZMediaOverlay(true)`](https://developer.android.com/reference/android/view/SurfaceView?hl=en#setZOrderMediaOverlay(boolean)) to pull FlutterView in front of another SurfaceView in their UI. That workaround won't help render a FlutterView transparently over most Android views (like `LinearLayout`, etc), but is a viable workaround for any other apps that are drawing over another `SurfaceView` with Flutter.

**Comment by phanirithvij** [comment]
@searchy2 how did you achieve transparency using `TransparencyMode`, do you mind providing an example?

**Comment by phanirithvij** [comment]
@mklim How to use the new API there are no examples in the [docs](https://api.flutter.dev/javadoc/io/flutter/embedding/android/FlutterView.TransparencyMode.html)?

**Comment by phanirithvij** [comment]
@searchy2 how were you able to use `TransparencyMode` do you mind providing a basic example?

**Comment by dannyvalentesonos** [comment]
Is there any plan to fix this? Or is there some workaround for this?

**Comment by phanirithvij** [comment]
> `customer: fun` has been able to work around this by calling [`setZMediaOverlay(true)`](https://developer.android.com/reference/android/view/SurfaceView?hl=en#setZOrderMediaOverlay(boolean)) to pull FlutterView in front of another SurfaceView in their UI. That workaround won't help render a FlutterView transparently over most Android views (like `LinearLayout`, etc), but is a viable workaround for any other apps that are drawing over another `SurfaceView` with Flutter.

@dannyvalentesonos have you tried this solution?

**Comment by dannyvalentesonos** [comment]
Thanks @phanirithvij . I'll have to look at that further when I have a chance, but we're using the new embedding API. I'm not sure we have access to the FlutterView easily, but we'll take a look.

**Comment by matthew-carroll** [comment]
@dannyvalentesonos to test the hypothesis, you can probably crawl the `View` hierarchy from `FlutterActivity` and invoke the necessary method on `FlutterSurfaceView`. If that solves the problem then we can look into adding some kind of public API for this setting, or perhaps we need to automatically invoke that method when talkback activates...

Also, I can't remember, but did you consider using the `TextureView` variant? It has a performance cost, but I don't think it suffers from this bug.

**Comment by dannyvalentesonos** [comment]
Thanks @matthew-carroll . I'll try that once I'm back in 2 weeks after the holidays, and I'll let you know. If it does work, I would like to see this kind of thing be called automatically when talkback activates, although it looks like it has to be called before the view is attached to the window... Is there a reason this isn't just always called? Is there a performance cost?

As for your question about the `TextureView` variant, I'm nervous about having any type of performance hit, seeing we already have some situations where building certain trees for the first time sometimes drops frames, and I'd be worried about making that worse.

**Comment by dnfield** [comment]
It seems like this is a bug in talkback - have we tried making a simple Android app reproduction of this?

**Comment by mklim** [comment]
@dnfield yeah, this is an issue with Talkback/Android. It's caused by SurfaceView's drawing "on top" of everything like described in [this comment](https://github.com/flutter/flutter/issues/37025#issuecomment-515584715), and it can be reproduced with just a simple Android app.

The true fix/workaround here is probably to switch to using a TextureView, like @matthew-carroll mentioned. The TextureView isn't drawn "on top" of the window and doesn't cover the Talkback highlights like the SurfaceView one does, so it doesn't have this bug. I think this is an inherent limitation of Talkback/SurfaceViews and can't really be meaningfully worked around with the v1 embedding. There's the `setZMediaOverlay` call I mentioned above, but that only works if you're trying to make the FlutterView transparent over another SurfaceView, and doesn't work if your'e trying to make it transparent over a typical Android UI. I think the easiest way to do that would be to switch to the v2 embedding since it's provided as a config option when setting up a FlutterView there.

> Is there a reason this isn't just always called? Is there a performance cost?

@dannyvalentesonos yeah, there's a performance cost to using TextureViews and so that's the main reason why we default to SurfaceViews. Basically TextureViews composite with everything nicely in the Android UI and work well for effects like this, but because they're compositing with everything else they can be much slower. [The AOSP docs](https://source.android.com/devices/graphics/arch-tv#surface_or_texture) have some good information on the differences. 

I think the final step for closing out this issue would be actually deleting the deprecated methods in the v1 embedding that lead to this buggy behavior.

**Comment by dannyvalentesonos** [comment]
@mklim Thanks for the response. I'm a little confused about your comment though. We are using the v2 embedding and still seeing this issue. Is there something we can do to fix this on our end, seeing we are using the v2 embedding?

**Comment by mklim** [comment]
No problem @dannyvalentesonos, sorry for being unclear. I am a little confused that you're seeing this bug with the v2 embedding because `enableTransparentBackground` shouldn't be an API that's available there. Are you still calling that API somehow or is this issue surfacing another way for you?

The v2 embedding still uses a SurfaceView by default, so on top of using it you also need to configure it to use a TextureView instead. The way to do that is in the `Activity` that extends `FlutterActivity`, override [`getBackgroundMode`](https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/android/FlutterActivity.java#L835) to always return `FlutterView.TransparencyMode.transparent`. That value is used in `getRenderMode` and eventually in [`FlutterView#init`](https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/android/FlutterView.java#L202) to set the underlying View to be a `TextureView` instead of a `SurfaceView`. As long as you're returning `TransparencyMode.transparent` from `getBackgroundMode` you should already be using a `TextureView`.

If you're doing that and still seeing this issue it's a new issue and not an understood one. Would appreciate any info you have on how to reproduce it, thanks!

**Comment by dannyvalentesonos** [comment]
Thanks @mklim .
I am not extending `FlutterActivity`.
Instead, I create a Flutter Fragment like so:
```
flutterFragment = FlutterFragment.withCachedEngine("my_engine")
                    .renderMode(FlutterView.RenderMode.surface)
                    .transparencyMode(FlutterView.TransparencyMode.transparent)
                    .build();
```
With this, I am not seeing the TalkBack borders being drawn. I can sometimes see just the top edges of the TalkBack corners around the outer most container which is uses Rounded corners as a decoration. I'm wondering if I have something more fundamentally wrong within my Flutter code?

**Comment by mklim** [comment]
Got it, thanks for the detailed info!

The problem line is this one:

```java
                    .renderMode(FlutterView.RenderMode.surface)
```

This is telling the FlutterEngine to use a `SurfaceView` instead of a `TextureView` while also being transparent, so it ends up covering Talkback like the v1 embedding would when `enableTransparentBackground` was called. If you replace it with: `.renderMode(FlutterView.RenderMode.texture)` the engine will switch to using a `TextureView` instead, so you shouldn't still see this issue. 

@matthew-carroll maybe the v2 embedding should log a warning or error referring to this bug when it's configured to be a SurfaceView that's also transparent, and suggest either turning off transparency or switching to `RenderMode.texture` instead?

**Comment by dannyvalentesonos** [comment]
Ok, thanks. However, switching to a TextureView was the concern I had around performance. I'd rather not have a transparent background while TalkBack is on, or maybe force TextureView while TalkBack is on, seeing in this case, not having a transparent background at all, or having slower performance would not be as much of an issue when TalkBack is on.
I'll probably implement that on my side then.
I wonder if Android recreates the activities like it does for Configuration Changes when Talkback is toggled on while the app is already running...

**Comment by matthew-carroll** [comment]
@mklim, looking back at this, what do you think the action items might be to consider this issue resolved? At this point we know that talkback is incompatible with `SurfaceView`s that sit atop the `Window`. By default the v2 embedding chooses `TextureView` when transparency is desired, though that can be overridden if the developer chooses. Is there anything else that Flutter can do?

**Comment by mklim** [comment]
I think with the v2 embedding updating a `TextureView` to a default and rolling to stable this can be closed. We could also delete the deprecated method in the v1 embedding, but I don't think that's critical.

**Comment by dnfield** [comment]
We should at least update the deprecation text to indicate how to do this nowdays.

**Comment by dnfield** [comment]
It currently just reads as deprecated with no work around.

**Comment by mklim** [comment]
Good point, I'll patch it to point to the updated flow.

**Comment by dannyvalentesonos** [comment]
Here's an interesting thing. I had locked my phone, and now when i came to unlock it, the talkback borders started drawing. So this tells me it's definitely possible to make work, however, it doesn't work when first launching the FlutterView.
@dnfield Just tagging you here to make sure we can get the right people to respond. Should I log a separate bug? Is a RenderMode.surface supposed to work with TransparancyMode.opaque? That is my understanding.

**Comment by dnfield** [comment]
There's some strange set of interleaved issues related to TalkBack, Android, and the Flutter embedding that makes all of this a little confusing.

It'd be helpful if you filed a new issue with your code to reproduce.  This issue was closed because the deprecation notice was updated with what was meant to be clearer instructions.
